content/blog/2019-10-23-register-allocation/index.md

```diff
@@ -0,0 +1,142 @@
+- `print <N args>;` has N memory access operations, N being the number of arguments (N loads)
+
+As for accessing register values (the case where special variables represented registers will not be counted as a memory access):
+
+- `r1: int = const 2;` has 0 memory access operations
+- `a: int = id r2;` has 1 memory access operations (store a)
+- `c:int = add a r2;` has 2 memory access operations (load a, store c)
+
+We compare the total number of memory access operations for the same program with and without register allocation. 
+
+It was identified during evaluation that this implementation of counting doesn't support the `ret` instruction. We hope to fix this issue shortly. We have tested this interpreter version with some hand written register allocated code.
+
+### Simulated load and store counts
+
+We simulated the baseline and graph coloring implementation using Brili to test for two things,
+1. To make sure that the functionality hasn't been affected by our optimization
+2. To measure memory access operations as a proxy for performance
+
+The following tables provides the memory acccess count in the form of `loads/stores`. These tests are run assuming for 4 registers.
+
+Basline and graph coloring tests are an average from multiple simulation runs. This turned out to be useful for the graph coloring technique, as the register assignment can change.
+
+| Test   | baseline | naïve | graph coloring | handtune |
+|--------|----------|-------|----------------|----------|
+| br     | 3/4      |       | 0/0            | 0/0      |
+| if     | 14/11    |       | 1/1            | 1/1      |
+| alloc1 | 6/5      | 0/0   | 0/0            | 0/0      |
+| alloc2 | 20/17    | 15/26 | 4/3            | 0/0      |
+| alloc3 | 20/16    | 17/27 | 8/4            | 2/2      |
+
+This first table evaluates several handcrafted tests to test whether the grach coloring implementation can handle all instructions and where does it lie in terms of performance in a space from inefficient baseline to optimized manual handtuning. As described in the implementation section, the implementation does not achieve optimal coloring, but manages to reduce memory accesses compared to other techniques.
+
+| Benchmark| baseline  | naïve     | graph coloring |
+|---------|-----------|-----------|----------------|
+| loop    | 53/24     |           | 21/2           |
+| matmul  | 2048/1216 | 1280/1980 | 1008/126       |
+| dotprod | 257/145   | 128/205   | 169/78         |
+| fib     | 103/56    |           | 32/12          |
+| fact    | 33/14     |           | 3/0.3           |
+| polymul | 929/509   | 711/767   | 869/505        |
+
+These benchmark tests stress test the implementations and provide interesting results. As seen from the tables, graph coloring technique manages to consistently have comparable or less accesses than the other two techniques. It also seem to compromise some load operations to avoid store operations.
```

Comment:
```suggestion
These benchmark tests stress test the implementations and provide interesting results. As seen from the tables, the graph coloring technique manages to consistently have comparable or less accesses than the other two techniques. It also seems to compromise some load operations to avoid store operations.
```